### ISSUE TITLE
[FlatList] Scrolling issues with FlatList when rows are variable height

### ISSUE THREAD

**Issue by giladno** [facebook/react-native#13727]
I'm using a `FlatList` where each row can be of different height (and may contain a mix of both text and zero or more images from a remote server).

I cannot use `getItemLayout` because I don't know the height of each row (nor the previous ones) to be able to calculate.

The problem I'm facing is that I cannot scroll to the end of the list (it jumps back few rows when I try) and I'm having issues when trying to use `scrollToIndex` (I'm guessing due to the fact I'm missing getItemLayout).

I wrote a sample project to demonstrate the problem (simply try to scroll to the end with your finger - it won't allow you)

```
import React, { Component } from 'react';
import { AppRegistry, StyleSheet, Text, View, Image, FlatList } from 'react-native';
import autobind from 'autobind-decorator';

const items = count => [...Array(count)].map((v, i) => ({
    key: i,
    index: i,
    image: 'https://dummyimage.com/600x' + (((i % 4) + 1) * 50) + '/000/fff',
}));

class RemoteImage extends Component {
    constructor(props) {
        super(props);
        this.state = {
            style: { flex: 1, height: 0 },
        };
    }

    componentDidMount() {
        Image.getSize(this.props.src, (width, height) => {
            this.image = { width, height };
            this.onLayout();
        });
    }

    @autobind
    onLayout(event) {
        if (event) {
            this.layout = {
                width: event.nativeEvent.layout.width,
                height: event.nativeEvent.layout.height,
            };
        }
        if (!this.layout || !this.image || !this.image.width)
            return;
        this.setState({
            style: {
                flex: 1,
                height: Math.min(this.image.height,
                    Math.floor(this.layout.width * this.image.height / this.image.width)),
            },
        });
    }
    render() {
        return (
            <Image
                onLayout={this.onLayout}
                source={{ uri: this.props.src }}
                style={this.state.style}
                resizeMode='contain'
            />
        );
    }
}

class Row extends Component {
    @autobind
    onLayout({ nativeEvent }) {
        let { index, item, onItemLayout } = this.props;
        let height = Math.max(nativeEvent.layout.height, item.height || 0);
        if (height != item.height)
            onItemLayout(index, { height });
    }

    render() {
        let { index, image } = this.props.item;
        return (
            <View style={[styles.row, this.props.style]}>
                <Text>Header {index}</Text>
                <RemoteImage src = { image } />
                <Text>Footer {index}</Text>
            </View>
        );
    }
}

export default class FlatListTest extends Component {
    constructor(props) {
        super(props);
        this.state = { items: items(50) };
    }

    @autobind
    renderItem({ item, index }) {
        return <Row
        item={item}
        style={index&1 && styles.row_alternate || null}
        onItemLayout={this.onItemLayout}
        />;
    }

    @autobind
    onItemLayout(index, props) {
        let items = [...this.state.items];
        let item = { ...items[index], ...props };
        items[index] = { ...item, key: [item.height, item.index].join('_') };
        this.setState({ items });
    }

    render() {
        return (
            <FlatList
                    ref={ref => this.list = ref}
                    data={this.state.items}
                    renderItem={this.renderItem}
                />
        );
    }
}

const styles = StyleSheet.create({
    row: {
        padding: 5,
    },
    row_alternate: {
        backgroundColor: '#bbbbbb',
    },
});

AppRegistry.registerComponent('FlatListTest', () => FlatListTest);
```

**Comment by MrChe** [comment]
@giladno why you use @autobind? maybe `=>` its very simple js binding!

**Comment by giladno** [comment]
@MrChe Arrow functions are bad for performances: https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-no-bind.md

**Comment by ugiacoman** [comment]
@giladno I'm facing the same issue. Any fixes? 

Example: https://gfycat.com/SaltyBitterGermanspitz

**Comment by MrChe** [comment]
@giladno  Very interesting information! Thk)

**Comment by giladno** [comment]
@ugiacoman no fix for me yet. I want to try the new 0.44 version to compare with.

**Comment by ugiacoman** [comment]
@giladno v0.44 Partially fixed it for me :)

**Comment by giladno** [comment]
@ugiacoman please explain "partially" :)

**Comment by ugiacoman** [comment]
@giladno So the FlatList correctly calculates the viewport so I can scroll all the way down. However if I add another component within the same Flexbox, it simply pushes it down and I end up not being able to scroll all the way. Now that I think about it, that may not be a bug. I need to do more testing to figure out the precise requirements so that I can ensure it renders correctly.

**Comment by ariona** [comment]
Hi @giladno any update for this? I am kinda facing the same problem, the current solution is to init the `initialNumToRender` to number of the item index i want to reach. and scrollToIndex when componentDidmounted.  But it's kinda slow

**Comment by giladno** [comment]
@ariona Sadly, no solution yet for me :(

**Comment by Ashoat** [comment]
I have this problem on a project of mine. My solution has been to dynamically calculate the height of each cell as a function of the data it contains, and to include the height information in the data blob I pass to the list so that I can implement `getItemLayout`. I block the rendering of the list on height calculation for literally every cell, which may be prohibitive for some folks depending on list length. The hardest part has been calculating the height of text, since that can depend on font, word wrapping, etc. On iOS native this can be achieved via the `sizeWithAttributes:` method, but there's no equivalent in React Native, so I pulled together a higher-order component that invisibly renders Text with a given style to the screen in batches, waits for `onLayout` to get the height, and then returns the text height via a callback prop.

**Comment by giladno** [comment]
@Ashoat Could you please share the code where you render the component and calculate its height? Thanks.

**Comment by Ashoat** [comment]
@gildano: np, here you go - https://gist.github.com/Ashoat/c283a40b052ae1e4571e944abcef0fd1

It's a React component called TextHeightMeasurer. You hand it:
- `textToMeasure` - an array of strings whose heights you want to measure
- `allHeightsMeasuredCallback` - a callback to call when measurement is completed
- `style` - the font (size, style, decoration, etc.) and content width should be the same as it will be when rendered in the list

Besides text, it should be possible to identify the height of everything else programmatically.

**Comment by giladno** [comment]
@Ashoat much appreciated buddy!!

**Comment by giladno** [comment]
@Ashoat Could you please contact me in private (check my profile page)? I have a consulting job for you.

**Comment by parkerproject** [comment]
@giladno v0.44 doesn't fix it, I just updated and still having the same issue.

**Comment by giladno** [comment]
@parkerproject Yeah, that's pretty bad. I'm willing to pay for a bug fix :(

**Comment by ippa** [comment]
Didn't have the exact same issue as you did, but I did have the same kind of data (images + random lengths of text).

When I switched to FlatList i ran into various perf-problems I couldn't get out of. Switching back to ListView solved it for me. Too bad, since the FlatList api is so much nicer.

**Comment by giladno** [comment]
@ippa Thanks, I was originally using ListView and it was ok, the problem is that it doesn't support jump to item, which is crucial in my case.

**Comment by lprhodes** [comment]
@Ashoat TextHeightMeasurer looks like it'll be hand thanks! How do you use it within an app though? Is it meant to replace a <Text> component or do you iterate through records and render the TextHeightMeasurer component off screen then display the list once the measurements are ready?

**Comment by Ashoat** [comment]
It's sorta like StatusBar, it doesn't render anything so you can place it anywhere. I place it in the same View my FlatList is in.

The version in the gist I linked above probably has a bug or two, here's the latest version: https://github.com/Ashoat/squadcal/blob/master/native/text-height-measurer.react.js

**Comment by lprhodes** [comment]
Nice, thanks.

I ended up using onLayout then storing the dimensions of each cell.

I then used onViewableItemsChanged to determine whether a cell should be visible or not and stored this information too.

Finally, the render method checks whether the cell should be visible or not. If not it returns an empty view with the previously stored version (otherwise cells would jump around). 

Additionally onViewableItemsChanged works out the direction that the scroll view is moving based on which cells are being removed. It then adds make sure there's always X cells visible before of the first cell and X cells visible after the last cell.

Doing this means I can now scroll through a lot more cells and the user only sees the blank cells in rare cases (a really low res image could be used to help with design).

I'll release it as a component next week some time.

**Comment by sahrens** [comment]
@ippa what perf issues did you have migrating off ListView?

**Comment by sahrens** [comment]
@giladno: does FlatList have any more problems than ListView for you, or is it mostly just the scrollToIndex with variable height problem?

**Comment by parkerproject** [comment]
This issue is more to do with the styling. My approach was to add footer component, though it might not be the optimal solution, it gets the job done for now
`ListFooterComponent={() => (
                  <View style={{ height: 150 }} />
                )}`

You can adjust the height to meet you needs

**Comment by sahrens** [comment]
I'm guessing there is an issue with the layout of your screen that's not FlatList specific. Try mucking with the style prop. A complete code snippet for the whole page might help others identify the bug.

**Comment by OrganicCat** [comment]
This may be a dumb question, but is there a reason you're not using "scrollToEnd()"?

I had a pretty big issue with this until I attached that. I was adding dynamically sized components, and it wasn't going all the way to the bottom. Note, that it doesn't fire properly under "componentDidUpdate()" so I created my own method which fires it:

```
scrollToBottom() {
    // Settimeout needed here to wait for the page module to be added so scroll works properly
    setTimeout(this.refs.listView.scrollToEnd, 0)
}
```

I'm not sure if this is NEEDED or I'm just missing something, but it's far easier than calculating each individual component size (unless you need to go to a specific component, not the bottom).

**Comment by Symous** [comment]
+1 I have met the similar issuse . any progress here ?

**Comment by DrPandemic** [comment]
@OrganicCat when do you call your `scrollToBottom`? When I try to call `scrollToEnd` too quickly (something faster than 1 second after the constructor) it doesn't scroll.

**Comment by OrganicCat** [comment]
```
  componentDidUpdate() {
     this.scrollToBottom()
  }
```

Also, I changed my timeout to 100 miliseconds.

**Comment by hramos** [comment]
Hi there! This issue is being closed because it has been inactive for a while. Maybe the issue has been fixed in a recent release, or perhaps it is not affecting a lot of people. Either way, we're automatically closing issues after a period of inactivity. Please do not take it personally!

If you think this issue should definitely remain open, please let us know. The following information is helpful when it comes to determining if the issue should be re-opened:

- Does the issue still reproduce on the latest release candidate? Post a comment with the version you tested.
- If so, is there any information missing from the bug report? Post a comment with all the information required by the [issue template](https://github.com/facebook/react-native/blob/master/.github/ISSUE_TEMPLATE.md).
- Is there a pull request that addresses this issue? Post a comment with the PR number so we can follow up.

If you would like to work on a patch to fix the issue, *contributions are very welcome*! Read through the [contribution guide](http://facebook.github.io/react-native/docs/contributing.html), and feel free to hop into [#react-native](https://discordapp.com/invite/0ZcbPKXt5bZjGY5n) if you need help planning your contribution.

**Comment by AnisDerbel** [comment]
you should not close the issue unless it's fixed , I'm having the same issue , I'm using a custom TreeView based on Flatlists , when you click on item the sub item will be visible then the height will be dynamic and the scrollToIndex doesn't work properly!

**Comment by mannol** [comment]
This issue still occurs to me on 0.47...

**Comment by dotansimha** [comment]
Same with 0.46... :(

**Comment by dotansimha** [comment]
For me, using `initialNumToRender` with `<FlatList ... style={{ flexGrow: 1}} />` works with 0.46

**Comment by eballeste** [comment]
@lprhodes would love too see your `onViewableItemsChanged` logic!

I'm currently experiencing jumps due to objects reverting to their collapsed state once they are off screen  (via `onViewableItemsChanged`) after the fact that a scrollToIndex was triggered by a user expanding some other, viewable, collapsed item.

**Comment by eballeste** [comment]
nevermind fixed the jumping by slowing down the collapsing item animation and setting
`viewabilityConfig={{ viewAreaCoveragePercentThreshold: 35 }}` 
on the flatlist so that expanded items start collapsing while partially off the screen as opposed to completely off screen, now it looks way better.

**Comment by victorbadila** [comment]
same issue, and this seems to be the standard behaviour. in the example page https://facebook.github.io/react-native/docs/using-a-listview.html if you add a lot of items to the FlatList example you will see that scrolling doesn't work properly, you scroll and then it gets back up to the beginning.
@hramos it's not nice to see you close a lot of react-native related issues due to inactivity, what's the point of this? at least have the documentation up to date, if this is the standard behaviour it should say so in the docs.

**Comment by sahrens** [comment]
Sorry, looking at this again I see what the issue is (although it doesn't seem to match a lot of the "me too" comments so I wonder the original issue was was edited? Sure would be nice if github let you [see edit history](https://github.com/isaacs/github/issues/954)).

In any case, it sounds like you're hitting the first caveat described in the docs. Quoting from https://facebook.github.io/react-native/docs/flatlist.html:

> Internal state is not preserved when content scrolls out of the render window. Make sure all your data is captured in the item data or external stores like Flux, Redux, or Relay.

So when your items scroll out of the render window, they lose their state which included the height from the async `getImage` callback (or onLayout). When they are back in the window (note fast scrolling will shrink the window and then it will try to grow in both directions once scrolling stops) they will render their images with zero height at first, then get the height again, and then adjust the height and push all the items below them down.

There are many ways to get around this. One would be a simple synchronous cache of src -> height that you would read from when initializing your `state`. But you already did the work of storing the height in the state of the `FlatListTest` component, you just never use it - you could pass `this.props.item.height` back in as a prop to your `RemoteImage` component and use that to initialize the state as well.

In conclusion, I don't see anything here that's not a documented caveat with a clear workaround, so this will remain closed as a won't fix. If you disagree, make a clear case or open a separate issue (and feel free to tag me on it).

**Comment by sahrens** [comment]
I think @eballeste is probably having that render window issue as well.

BUT, as for the bottom content not being reachable, I still think it probably has to do with the layout of the rest of the app. @victorbadila mentioned it's reproable in the demo app, but it also repros with a plain ScrollView and ListView, not just FlatList. I'm guessing that something is messed up a shared component like the navigation library, or something like that. The proper fix will depend on the rest of the app, but you can also just add some marginBottom.

**Comment by eballeste** [comment]
@sahrens, nope! I'm good, I found the `viewOffset` parameter for the `FlatList.scrollToIndex` method. My feed feels really good.

**Comment by sahrens** [comment]
Great to hear!

**Comment by Knight704** [comment]
I use onLayout callback and inside it scrollToEnd work as expected, without getItemLayout implementation

**Comment by ariona** [comment]
I use that method too but when i want to navigate to 200th item it has long delay.

For now i use react-native-recyclerview-list but it only available for android.

**Comment by xavieramoros** [comment]
I'm having the same issue: a flatlist with variable height items and scrollToIndex not scrolling to the right position. 

But I noticed something interesting, the FlatList scrolls to the right position to then adjust to the wrong position:

Here is an example of the behaviour: 
![flatlistscrolling](https://user-images.githubusercontent.com/347030/32942996-fc1e1b44-cb8a-11e7-9fd4-36313c72490e.gif)
https://exp.host/@xavieramoros/snack-SyN4FJikz
source code: 
https://github.com/xavieramoros/flatlist-initialScrollIndexIssue/blob/master/App.js
Expo 22 - RN 0.49

If the FlatList is able to scroll (even if briefly) to the right index, couldn't we just find a way make it stay there?

I tried adding:
> viewabilityConfig={{ viewAreaCoveragePercentThreshold: 35 }}
as @eballeste suggested, didn't work for me.

My last option is to previously calculate the height of each item (what @Ashoat suggested), but it's hard since in my original project I'm rendering facebook posts, tweets, instagram posts, blogs posts all with variable text and images.

**Comment by eballeste** [comment]
@xavieramoros, that's what I'm doing, I save the heights of the rendered
rows in my redux state and use the `viewOffset` parameter of the
`scrollToIndex` method to compensate for that jump that you see.

The `viewabilityConfig.viewAreaCoveragePercentThreshold` prop I use to
figure out when should the row be considered to be out of screen.

On Nov 17, 2017 5:42 AM, "xavieramoros" <notifications@github.com> wrote:

I'm having the same issue: a flatlist with variable height items and
scrollToIndex not scrolling to the right position.

But I noticed something interesting, the FlatList scrolls to the right
position to then adjust to the wrong position:

Here is an example of the behaviour:
[image: flatlistscrolling]
<https://user-images.githubusercontent.com/347030/32942996-fc1e1b44-cb8a-11e7-9fd4-36313c72490e.gif>
https://exp.host/@xavieramoros/snack-SyN4FJikz
source code:
https://github.com/xavieramoros/flatlist-initialScrollIndexIssue/blob/
master/App.js
Expo 22 - RN 0.49

If the FlatList is able to scroll (even if briefly) to the right index,
couldn't we just find a way make it stay there?

I tried adding:

viewabilityConfig={{ viewAreaCoveragePercentThreshold: 35 }}
as @eballeste <https://github.com/eballeste> suggested, didn't work for me.

My last option is to previously calculate the height of each item (what
@Ashoat <https://github.com/ashoat> suggested), but it's hard since in my
original project I'm rendering facebook posts, tweets, instagram posts,
blogs posts all with variable text and images.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/facebook/react-native/issues/13727#issuecomment-345208544>,
or mute the thread

**Comment by pcattori** [comment]
Solved this by specifying both `getItemLayout` and abandoning `initialScrollIndex` in favor of calling `scrollToOffset` within `onLayout`.

I had a list of ~300 elements with variable heights. I knew ahead of time that certain rows would be of certain heights, so i pre-computed each row height (for `length` property) and cumulative row height (for `offset` property) to use in `getItemLayout`. then i used `onLayout` to call `scrollToOffset`.

Was getting (1) blank screens, (2) jittery movement, (3) flashes of the correct UI suddenly disappearing etc... before this fix.

```
const cumulativeRowHeights = [
  0,
  ...rowHeights.reduce(function(soFar, current) {
    const next = (soFar[soFar.length - 1] || 0) + current;
    soFar.push(next);
    return soFar;
  }, [])
].slice(0, data.length);

getItemLayout = (data, index) => {
        const length = rowHeights[index];
        const offset = cumulativeRowHeights[index];
        return {length, offset, index};
}

onLayout = () => this.listRef.scrollToOffset(cumulativeRowHeights[desiredIndex])
```

**Comment by xavieramoros** [comment]
Sorry for the late reply, what you mention @eballeste is very similar to what @pcattori uses. It's easy  to calculate the offset when you know the height of the elements. I've updated the snack with both approaches (scrollToIndex and scrollToOffset) working correctly :
https://exp.host/@xavieramoros/snack-SyN4FJikz

But in my (real) case, my problem is that calculating the height of the elements is quite difficult . I was trying to find another way but I'll have to stick to these two for now.

Thanks both!

**Comment by RoarRain** [comment]
Hi @xavieramoros Can I look at the https://exp.host/@xavieramoros/snack-SyN4FJikz  source code？ thx！

**Comment by wellyshen** [comment]
I have the same issue.

**Comment by Sir-hennihau** [comment]
Any fixes on this yet?

**Comment by victorbadila** [comment]
from what I have seen it sometimes happens and sometimes does not, even though in the examples where I use it the code is really the same. once you get it to "work" in a specific example it will continue to work forever, so the behaviour seems to be pretty much deterministic. not sure what triggers it to not work in other examples.

**Comment by Sir-hennihau** [comment]
> not sure what triggers it to not work in other examples.

Unfortunate if you can not get it running even once :D.

**Comment by xavieramoros** [comment]
@RoarRain the source code is here: https://github.com/xavieramoros/flatlist-initialScrollIndexIssue

**Comment by wenkangzhou** [comment]
I have the same issue.

**Comment by afshin-hoseini** [comment]
This worked for me:

**Comment by tvelev92** [comment]
Still haven't been able to find a solution

**Comment by anasidrissi** [comment]
thi smy be usefull if any one have problem with scrolling
```
<FlatList numColumns={3} columnWrapperStyle={{
                flex:1,
                flexDirection:'row',
                flexWrap:'wrap',
            }}
```

**Comment by itome** [comment]
I have the same issue. https://github.com/godness84/react-native-recyclerview-list can solve this issue only for android.
But I still don't know how to implement this feature in iOS.

**Comment by MN767** [comment]
I did not find any way to use `getItemLayout`  when the rows have variable heights , So you can not use `initialScrollIndex` . 

But I have a solution that may be a bit slow:
You can use `scrollToIndex` ,  but when your item is rendered . So you need `initialNumToRender` . 
You have to wait for the item to be rendered and after use  `scrollToIndex` so you can not use `scrollToIndex` in `componentDidMount` .

The only solution that comes to my mind is using `scrollToIndex` in `onViewableItemsChanged` . Take note of the example below :
In this example, we want to go to item `this.props.index` as soon as this component is run

```
constructor(props){

        this.goToIndex = true;

    }

render() {
        return (

                <FlatList
                    ref={component => {this.myFlatList = component;}}
                    data={data}
                    renderItem={({item})=>this._renderItem(item)}
                    keyExtractor={(item,index)=>index.toString()}
                    initialNumToRender={this.props.index}
                    onViewableItemsChanged={({ viewableItems }) => {
                        if (this.goToIndex){
                            this.goToIndex = false;
                            setTimeout(() => { this.myFlatList.scrollToIndex({index:this.props.index}); }, 10);
                        }
                    }}
                />

        );
    }

```
